Cleanup in string preprocessing

romainbrenguier opened the pull request:
This is based on https://github.com/diffblue/cbmc/pull/2374
The goal is to ensure location is set for each created symbol.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their

romainbrenguier committed:
Add missing const's and move's

Add const keyword at the declaration of variables that will not get
modified.
In some case, we inline them instead to avoid declaration, when
declaration doesn't add value.
In other cases, we don't make the variable const but use move where it
is used.

fixup! Add missing const's

romainbrenguier committed:
Add and use fresh_java_symbol function

This prevents forgetting that the basename of the symbol should be the
function identifier.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2b45d8037e29081e7bc3271ab4fb79491fc790da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88131981

smowton commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -436,3 +437,20 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+  const source_locationt &source_location,
+  symbol_table_baset &symbol_table)
+{
+  const std::string name_prefix = id2string(source_location.get_function());
```
Check you got a real source loc in this case (or handle the no-source-loc case somehow)

tautschnig replied:
I believe @peterschrammel has comment on this (or a similar case) elsewhere: this seems actually wrong, instead the internal identifier of the function should be passed in.

romainbrenguier replied:
Yes the conversation was on https://github.com/diffblue/cbmc/pull/2374, I will copy it here so that it doesn't get lost now that 2374 is closed, and address it when I have time to get back to this PR.

> tautschnig 13 days ago  Collaborator
> I strongly believe that the information in the source_location is a string aimed at the user, and should relate to whatever is found in the source. The actual identifier of the function (the key in goto_functionst::function_map) used internally may be entirely different. I don't know which one is required here.

> peterschrammel 13 days ago  Member
Here, something is required that is unique to a goto function. Reason: adding a function to a goto model must not affect the identifiers of another function. This was the source of many goto-diff bugs in the past.

> tautschnig 13 days ago  Collaborator
Then a function_id must be passed in. In some call sites in this PR this would be symbol.name, others require further digging.

smowton reviewed (approved):
Looks good, one thing to check re: get-fresh-java-symbol, and also please check you intended to include a models lib bump in this PR

romainbrenguier committed:
Add function_id parameter to decl_string_expr

This is in preparation for giving function_id as an argument to the
function creating fresh java symbols.

romainbrenguier committed:
Add function_id parameter to add_pointer_to_array_association

This is in preparation for giving the function name to the function
generating fresh java symbols.

romainbrenguier committed:
Add function_id parameter to add_array_to_length_association

This is in preparation to giving the function_id to the function
generating fresh java symbols.

romainbrenguier committed:
Add function_id parameter to convert_exprt_to_string_exprt

This is in preparation to giving the function_id to the function
generating fresh java symbols.

romainbrenguier committed:
Add function_id parameter to fresh_string

This is in preparation to giving the function_id to the function
generating fresh java symbols.

romainbrenguier committed:
Add function_id parameter to process_operands

This is in preparation for adding the function_id to the function
generating fresh java symbols.

romainbrenguier committed:
Apply clang-format suggestions

No functional changes.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8de539f0c0bc6157e4fcde017045ed947cfb47b6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99921522

romainbrenguier commented:
@peterschrammel this is now rebased and ready for review

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -278,7 +278,7 @@ exprt::operandst java_string_library_preprocesst::process_parameters(
   {
-    symbol_exprt sym(p.get_identifier(), p.type());
+    const symbol_exprt sym(p.get_identifier(), p.type());
     ops.push_back(sym);
```
I'd go the opposite direction: not use `const` and use `ops.push_back(std::move(sym));` or even just do `ops.emplace_back(symbol_exprt(p.get_identifier(), p.type()));` - `sym` doesn't add much value anyway.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -451,21 +451,21 @@ refined_string_exprt java_string_library_preprocesst::replace_char_array(
   // array_data is array_pointer-> data
   const exprt array_data = get_data(array, symbol_table);
-  symbolt sym_char_array = get_fresh_aux_symbol(
+  const symbolt sym_char_array = get_fresh_aux_symbol(
```
Use a `const symbolt &` - no need for a by-value copy.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -502,25 +502,26 @@ refined_string_exprt java_string_library_preprocesst::decl_string_expr(
   code_blockt &code)
 {
-  symbolt sym_length = get_fresh_aux_symbol(
+  const symbolt sym_length = get_fresh_aux_symbol(
```
`const symbolt &`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -502,25 +502,26 @@ refined_string_exprt java_string_library_preprocesst::decl_string_expr(
-  symbolt sym_content = get_fresh_aux_symbol(
+  const symbol_exprt length_field = sym_length.symbol_expr();
+  const pointer_typet array_type = pointer_type(java_char_type());
+  const symbolt sym_content = get_fresh_aux_symbol(
```
`const symbolt &`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -602,7 +603,7 @@ static codet code_assign_function_application(
   symbol_table_baset &symbol_table)
 {
-  exprt fun_app=make_function_application(
+  const exprt fun_app = make_function_application(
```
I'd not use `const`, and would use `std::move` below.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -622,8 +623,8 @@ codet java_string_library_preprocesst::code_return_function_application(
-    function_name, arguments, type, symbol_table);
+  const exprt fun_app =
+    make_function_application(function_name, arguments, type, symbol_table);
   return code_returnt(fun_app);
```
Inline the `make_function_application` to avoid unnecessary temporaries.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1107,10 +1108,11 @@ code_blockt java_string_library_preprocesst::make_init_function_from_call(
   // Processing parameters
-  exprt::operandst args=process_parameters(params, loc, symbol_table, code);
+  const exprt::operandst args =
+    process_parameters(params, loc, symbol_table, code);
```
Please `std::move` (and thus don't `const`) this, that's an unnecessary copying of a vector.

romainbrenguier replied:
I think the move constructor would already be used here since the right-hand-side is movable, and the function where it's used takes the argument by reference, so there is not point using move there.

tautschnig replied:
As far as I understand `args` is not implicitly moved here for two reasons: 1) it's a named object and 2) it's `const`. But then also my comment may have been unclear: the value will be moved into `args`, but I was asking for it also to be moved out of `args`.

romainbrenguier replied:
Ok I thought you wanted to move into args. Moving it out of args will have no effect here because the function `string_expr_of_function` takes a const ref.

tautschnig replied:
Ok, sorry, please disregard my comment in that case.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1137,7 +1139,7 @@ java_string_library_preprocesst::make_assign_and_return_function_from_call(
 {
   // This is similar to assign functions except we return a pointer to `this`
-  java_method_typet::parameterst params = type.parameters();
+  const java_method_typet::parameterst params = type.parameters();
```
Take a reference, don't copy.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1246,27 +1248,28 @@ java_string_library_preprocesst::get_primitive_value_of_object(
-  std::string aux_name="tmp_"+id2string(type_name);
-  symbolt symbol=get_fresh_aux_symbol(
+  const std::string aux_name = "tmp_" + id2string(type_name);
+  const symbolt symbol = get_fresh_aux_symbol(
```
`const symbolt &`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1246,27 +1248,28 @@ java_string_library_preprocesst::get_primitive_value_of_object(
       symbol_table.lookup(object_type->get_identifier()))
   {
-    const struct_typet &struct_type = to_struct_type(maybe_symbol->type);
+    const struct_typet struct_type = to_struct_type(maybe_symbol->type);
```
`const struct_typet &`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1246,27 +1248,28 @@ java_string_library_preprocesst::get_primitive_value_of_object(
+      const dereference_exprt deref(
         typecast_exprt(object, pointer_type), pointer_type.subtype());
-      member_exprt deref_value(deref, value_comp.get_name(), value_comp.type());
+      const member_exprt deref_value(
```
Use `std::move` below, don't `const`

romainbrenguier replied:
`code_assignt`'s constructor takes argument by const ref so `move` would have no effect here (though exprt should have constructors from movable arguments, that's outside the scope of this PR).

romainbrenguier replied:
`member_exprt` and `code_assignt`'s constructors take arguments by const ref so `move` would have no effect here

tautschnig replied:
True at the moment, but #3502 will eventually fix that. So whenever I have cleaned up #3502 it would be nice if code also were making use of it.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1246,27 +1248,28 @@ java_string_library_preprocesst::get_primitive_value_of_object(
     {
-      pointer_typet pointer_type=::pointer_type(struct_type);
-      dereference_exprt deref(
+      const pointer_typet pointer_type = ::pointer_type(struct_type);
```
While changing this, also get rid of `::`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1517,22 +1520,21 @@ code_blockt java_string_library_preprocesst::make_object_get_class_code(
-  symbolt class1_sym=get_fresh_aux_symbol(
+  const pointer_typet class_type =
+    java_reference_type(symbol_table.lookup_ref("java::java.lang.Class").type);
+  const symbolt class1_sym = get_fresh_aux_symbol(
```
`const symbolt &`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1591,8 +1593,8 @@ code_blockt java_string_library_preprocesst::make_function_from_call(
-  exprt::operandst args=process_parameters(
-    type.parameters(), loc, symbol_table, code);
+  const exprt::operandst args =
+    process_parameters(type.parameters(), loc, symbol_table, code);
```
Don't, make sure you can `std::move` that vector.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1624,15 +1626,15 @@ java_string_library_preprocesst::make_string_returning_function_from_call(
-  exprt::operandst arguments=process_parameters(
-    type.parameters(), loc, symbol_table, code);
+  const exprt::operandst arguments =
+    process_parameters(type.parameters(), loc, symbol_table, code);
```
Don't, make sure you `std::move` it.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1668,17 +1670,18 @@ code_blockt java_string_library_preprocesst::make_copy_string_code(
 
   // Assign the argument to string_expr
-  java_method_typet::parametert op = type.parameters()[0];
+  const java_method_typet::parametert op = type.parameters()[0];
```
Reference?

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1714,18 +1717,19 @@ code_blockt java_string_library_preprocesst::make_copy_constructor_code(
 
   // Assign argument to a string_expr
-  java_method_typet::parameterst params = type.parameters();
+  const java_method_typet::parameterst params = type.parameters();
```
Reference

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1753,10 +1757,10 @@ code_returnt java_string_library_preprocesst::make_string_length_code(
   (void)function_id;
 
-  java_method_typet::parameterst params = type.parameters();
+  const java_method_typet::parameterst params = type.parameters();
```
Reference

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -307,7 +307,7 @@ java_string_library_preprocesst::convert_exprt_to_string_exprt(
   PRECONDITION(implements_java_char_sequence_pointer(expr_to_process.type()));
   const refined_string_exprt string_expr =
-    decl_string_expr(loc, symbol_table, init_code);
+    decl_string_expr(loc, loc.get_function(), symbol_table, init_code);
```
This is most likely not the correct name, you'd want the name that is known to the symbol table. Really really really only use `source_locationt::get_function` if you are generating output to humans.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -494,11 +494,13 @@ symbol_exprt java_string_library_preprocesst::fresh_string(
 /// Add declaration of a refined string expr whose content and length are
 /// fresh symbols.
 /// \param loc: source location
+/// \param function_name: name of the function in which the stirng is defined
```
s/stirng/string/

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -540,7 +542,8 @@ refined_string_exprt java_string_library_preprocesst::make_nondet_string_expr(
   /// \todo refactor with initialize_nonddet_string_struct
-  const refined_string_exprt str = decl_string_expr(loc, symbol_table, code);
+  const refined_string_exprt str =
+    decl_string_expr(loc, loc.get_function(), symbol_table, code);
```
As above, I do question the use of `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -469,7 +469,12 @@ refined_string_exprt java_string_library_preprocesst::replace_char_array(
+    inf_array,
+    symbol_table,
+    loc,
+    loc.get_function(),
```
Don't use `get_function`, see other comment.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -335,7 +337,8 @@ exprt::operandst java_string_library_preprocesst::process_operands(
       ops.push_back(
-        convert_exprt_to_string_exprt(p, loc, symbol_table, init_code));
+        convert_exprt_to_string_exprt(
+          p, loc, symbol_table, loc.get_function(), init_code));
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -367,14 +370,16 @@ java_string_library_preprocesst::process_equals_function_operands(
   ops.push_back(
-    convert_exprt_to_string_exprt(op0, loc, symbol_table, init_code));
+    convert_exprt_to_string_exprt(
+      op0, loc, symbol_table, loc.get_function(), init_code));
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -367,14 +370,16 @@ java_string_library_preprocesst::process_equals_function_operands(
   ops.push_back(
-    convert_exprt_to_string_exprt(tcast, loc, symbol_table, init_code));
+    convert_exprt_to_string_exprt(
+      tcast, loc, symbol_table, loc.get_function(), init_code));
```
Don't use `get_function`

tautschnig commented on jbmc/unit/java_bytecode/java_string_library_preprocess/convert_exprt_to_string_exprt.cpp:
```diff
@@ -25,14 +25,15 @@ refined_string_exprt convert_exprt_to_string_exprt_unit_test(
 {
   return preprocess.convert_exprt_to_string_exprt(
-    deref, loc, symbol_table, init_code);
+    deref, loc, symbol_table, loc.get_function(), init_code);
```
Don't use `get_function`

tautschnig commented on jbmc/unit/java_bytecode/java_string_library_preprocess/convert_exprt_to_string_exprt.cpp:
```diff
@@ -25,14 +25,15 @@ refined_string_exprt convert_exprt_to_string_exprt_unit_test(
   GIVEN("A location, a string expression, and a symbol table")
   {
     source_locationt loc;
+    loc.set_function("function_name");
```
If this is necessary in a unit test, then most likely something is wrong.

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -322,23 +324,24 @@ java_string_library_preprocesst::convert_exprt_to_string_exprt(
 exprt::operandst java_string_library_preprocesst::process_operands(
   const exprt::operandst &operands,
   const source_locationt &loc,
+  const irep_idt &function_name,
```
Sometimes it's `function_name`, sometimes it's `function_id`. Can we only use the latter?

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_instrument.cpp:
```diff
@@ -112,14 +111,12 @@ code_ifthenelset java_bytecode_instrumentt::throw_exception(
+    exc_ptr_type,
+    "new_exception",
+    original_loc,
+    original_loc.get_function(),
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -251,12 +251,11 @@ symbol_exprt java_string_library_preprocesst::fresh_array(
-    "cprover_string_array",
     location,
-    ID_java,
+    location.get_function(),
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -463,8 +462,8 @@ refined_string_exprt java_string_library_preprocesst::replace_char_array(
-  const symbolt sym_char_array = get_fresh_aux_symbol(
-    array_data.type(), "char_array", "char_array", loc, ID_java, symbol_table);
+  const symbolt sym_char_array = fresh_java_symbol(
+    array_data.type(), "char_array", loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -766,8 +744,8 @@ void add_character_set_constraint(
-  const symbolt &return_sym = get_fresh_aux_symbol(
-    java_int_type(), "cnstr_added", "cnstr_added", loc, ID_java, symbol_table);
+  const symbolt &return_sym = fresh_java_symbol(
+    java_int_type(), "cnstr_added", loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1271,8 +1248,8 @@ java_string_library_preprocesst::get_primitive_value_of_object(
-  const symbolt symbol = get_fresh_aux_symbol(
-    value_type, aux_name, aux_name, loc, ID_java, symbol_table);
+  const symbolt symbol = fresh_java_symbol(
+    value_type, aux_name, loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1377,8 +1354,8 @@ struct_exprt java_string_library_preprocesst::make_argument_for_format(
-      symbolt field_symbol = get_fresh_aux_symbol(
-        type, id2string(function_id), tmp_name, loc, ID_java, symbol_table);
+      symbolt field_symbol = fresh_java_symbol(
+        type, tmp_name, loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1393,13 +1370,8 @@ struct_exprt java_string_library_preprocesst::make_argument_for_format(
-    ID_java,
-    symbol_table);
+  const symbolt object_symbol = fresh_java_symbol(
+    obj.type(), "tmp_format_obj", loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1544,8 +1516,8 @@ code_blockt java_string_library_preprocesst::make_object_get_class_code(
-  const symbolt class1_sym = get_fresh_aux_symbol(
-    class_type, "class_symbol", "class_symbol", loc, ID_java, symbol_table);
+  const symbolt class1_sym = fresh_java_symbol(
+    class_type, "class_symbol", loc, loc.get_function(), symbol_table);
```
Don't use `get_function`

tautschnig commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -19,6 +19,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <set>
 #include <unordered_set>
+#include <util/fresh_symbol.h>
```
Move to the `util` section above

romainbrenguier committed:
Add function_id parameter to replace_char_array

This is in preparation for giving the function name to the function
generating fresh java symbols.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6997a769418122c46fd5c68763566b39be155d71).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100205172

romainbrenguier committed:
Delete unused function fresh_array

This function is not used, there is no point in keeping it.

romainbrenguier committed:
Add function_id parameter to add_character_set_constraint

This is in preparation for using a new fresh_java_symbol function which
takes a function_id as parameter.

romainbrenguier committed:
Add function_id parameter to get_primitive_value_of_object

This is in preparation to giving the function_id to the function
generating fresh java symbols.

romainbrenguier committed:
Delete unused function process_operands

This is not used

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e81e174e8b9255f29a1f2f63b58b8295fdbbfd89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100241422

romainbrenguier commented:
@tautschnig I got rid of all the `get_function` calls

tautschnig reviewed (approved):
Thanks a lot for the large-scale cleanup and updates!
